quimb/tensor/optimize.py

```diff
@@ -147,10 +153,70 @@ def parse_network_to_backend(tn, tags, constant_tags, to_constant):
         if isinstance(data, qarray):
             data = data.A
 
-        # append the raw data but mark the corresponding tensor for reinsertion
+        return data
+
+    # handle simple case of every tensor being a variable
+    if not opt_tags:
+        for t in tn_ag:
+            # append the raw data but mark the corresponding tensor
+            # for reinsertion
+            data = get_tensor_data(t)
+            variables.append(data)
+            t.add_tag(variable_tag.format(len(variables) - 1))
+        return tn_ag, variables
+
+    # handle normally tagged tensors
+    for t in tn_ag.select(tags, 'any'):
+        # append the raw data but mark the corresponding tensor
+        # for reinsertion
+        data = get_tensor_data(t)
         variables.append(data)
         t.add_tag(variable_tag.format(len(variables) - 1))
 
+    # handle shared tags
+    for tag in shared_tags:
+
+        var_name = variable_tag.format(len(variables))
+        test_data = None
+
+        for t in tn_ag.select(tag):
+            data = get_tensor_data(t)
+
+            # detect that this tensor is already variable tagged and skip
+            # if it is
+            if any([variable_finder.match(tag) for tag in t.tags]):
```

Comment:
No need to change but thought I'd mention functions like `any` take a generator (i.e. you can remove the square brackets) - this also makes them more efficient since they lazily short circuit as soon as any True is found.

Reply:
Ah, that's good to know, thanks!